PocketKit/Sources/RecentSavesWidgetKit/RecentSaves/Provider/RecentSavesProvider.swift

```diff
@@ -0,0 +1,166 @@
+            // logged out
+            guard service.isLoggedIn else {
+                completion(RecentSavesEntry(date: Date(), contentType: .loggedOut))
+                return
+            }
+            let saves = service.getRecentSaves(limit: numberOfItems(for: context.family))
+            // empty result
+            guard !saves.isEmpty else {
+                completion(RecentSavesEntry(date: Date(), contentType: .empty))
+                return
+            }
+            Task {
+                let contentWithImages = await getContentWithImages(content: saves)
+                completion(RecentSavesEntry(date: Date(), contentType: .items(contentWithImages)))
+            }
+        } catch {
+            Log.capture(message: "Recent Saves widget: unable to initialize service - \(error)")
+            completion(RecentSavesEntry(date: Date(), contentType: .error))
+        }
+    }
+
+    func getTimeline(in context: Context, completion: @escaping (Timeline<RecentSavesEntry>) -> Void) {
+        do {
+            let service = try makeService()
+            // logged out
+            guard service.isLoggedIn else {
+                let entries = [RecentSavesEntry(date: Date(), contentType: .loggedOut)]
+                let timeline = Timeline(entries: entries, policy: .never)
+                completion(timeline)
+                return
+            }
+            let saves = service.getRecentSaves(limit: numberOfItems(for: context.family))
+            // empty result
+            guard !saves.isEmpty else {
+                let entries = [RecentSavesEntry(date: Date(), contentType: .empty)]
+                let timeline = Timeline(entries: entries, policy: .never)
+                completion(timeline)
+                return
+            }
+            Task {
+                let contentWithImages = await getContentWithImages(content: saves)
+                let entriesWithImages = [RecentSavesEntry(date: Date(), contentType: .items(contentWithImages))]
+                let timeline = Timeline(entries: entriesWithImages, policy: .never)
+                completion(timeline)
+            }
+        } catch {
+            Log.capture(message: "Recent Saves widget: unable to initialize service - \(error)")
+            let timeline = Timeline(entries: [RecentSavesEntry(date: Date(), contentType: .error)], policy: .never)
+            completion(timeline)
+        }
+    }
+}
+
+// MARK: read recent saves
+extension RecentSavesProvider {
+    /// Returns the number of recent saves to display for a given widget family
+    /// - Parameter widgetFamily: the given widget family
+    /// - Returns: the number of recent saves
+    private func numberOfItems(for widgetFamily: WidgetFamily) -> Int {
+        switch widgetFamily {
+        case .systemExtraLarge:
+            return 0
+        case .systemLarge:
+            return 4
+        case .systemMedium:
+            return 2
+        default:
+            return 1
```

Comment:
Is it just me or do these numbers seem random?

Reply:
They're not random: we currently support large and medium, the others are placeholders for future format support.